File: web/extensions/parts-of-speech.ts
// Parts-of-speech highlighting — an extension, not a feature of the editor.
//
// Like `typewriter.ts`, nothing in `web/src/` knows this exists. It reads the document
// text, decides what each word is, and pushes a decoration layer (DESIGN §5.3).
//
// It is the more demanding of the two showcase extensions, because its decorations
// depend on *language*, which the markdown parser has no concept of and never will.
//
// A caveat worth stating plainly rather than burying: the tagger below is a heuristic —
// a closed-class word list plus suffix rules. It is not a real part-of-speech tagger and
// it will be wrong on genuinely ambiguous words ("book a flight" tags `book` as a noun).
// The Apple build uses the operating system's `NLTagger` and is much better. What is
// being demonstrated here is the *plumbing* — that a feature this far outside markdown
// can drive the editor's decoration pipeline without the editor knowing about it.

import type { MarkdownEditor } from '../src/editor.js';

const DETERMINERS = new Set([
  'a', 'an', 'the', 'this', 'that', 'these', 'those', 'my', 'your', 'his', 'her',
  'its', 'our', 'their', 'some', 'any', 'no', 'every', 'each', 'either', 'neither',
]);

const PRONOUNS = new Set([
  'i', 'you', 'he', 'she', 'it', 'we', 'they', 'me', 'him', 'us', 'them', 'who',
  'whom', 'which', 'what', 'whose', 'someone', 'anyone', 'everyone', 'nothing',
]);

const PREPOSITIONS = new Set([
  'of', 'in', 'to', 'for', 'with', 'on', 'at', 'from', 'by', 'about', 'as', 'into',
  'like', 'through', 'after', 'over', 'between', 'out', 'against', 'during', 'without',
  'before', 'under', 'around', 'among', 'up', 'down', 'off', 'near', 'per',
]);

const CONJUNCTIONS = new Set([
  'and', 'but', 'or', 'nor', 'so', 'yet', 'because', 'although', 'though', 'while',
  'if', 'unless', 'until', 'whereas', 'than', 'that', 'when', 'where',
]);

// Auxiliaries and very common irregular verbs, which no suffix rule can reach.
const VERBS = new Set([
  'is', 'am', 'are', 'was', 'were', 'be', 'been', 'being', 'has', 'have', 'had',
  'do', 'does', 'did', 'will', 'would', 'can', 'could', 'shall', 'should', 'may',
  'might', 'must', 'get', 'got', 'go', 'goes', 'went', 'gone', 'make', 'made',
  'say', 'says', 'said', 'see', 'saw', 'seen', 'take', 'took', 'taken', 'come',
  'came', 'know', 'knew', 'known', 'give', 'gave', 'given', 'find', 'found',
  'think', 'thought', 'tell', 'told', 'become', 'became', 'leave', 'left', 'put',
  'keep', 'kept', 'let', 'begin', 'began', 'seem', 'help', 'show', 'shown',
  'hear', 'heard', 'run', 'ran', 'move', 'moved', 'live', 'bring', 'brought',
  'write', 'wrote', 'written', 'read', 'draw', 'drew', 'stay', 'stays',
]);

const ADJECTIVES = new Set([
  'good', 'new', 'first', 'last', 'long', 'great', 'little', 'own', 'other', 'old',
  'right', 'big', 'high', 'different', 'small', 'large', 'next', 'early', 'young',
  'important', 'few', 'public', 'bad', 'same', 'able', 'best', 'better', 'whole',
  'real', 'clean', 'fast', 'simple', 'plain', 'quick', 'brown', 'lazy',
]);

const ADVERBS = new Set([
  'not', 'also', 'very', 'often', 'however', 'too', 'usually', 'really', 'early',
  'never', 'always', 'sometimes', 'together', 'likely', 'simply', 'still', 'just',
  'already', 'only', 'even', 'again', 'here', 'there', 'now', 'then', 'once',
  'rather', 'quite', 'almost', 'instead', 'exactly', 'deliberately',
]);

/**
 * A coarse tag for one lowercase word, or null for "leave it alone".
 *
 * Order matters: closed classes are decided by membership, and only then do suffix
 * rules run. A suffix rule that fired first would call "the" an adjective.
 */
export function tagWord(word) {
  const w = word.toLowerCase();

  if (DETERMINERS.has(w) || PRONOUNS.has(w) || PREPOSITIONS.has(w) || CONJUNCTIONS.has(w)) {
    return null;
  }
  if (ADVERBS.has(w)) return 'adverb';
  if (VERBS.has(w)) return 'verb';
  if (ADJECTIVES.has(w)) return 'adjective';

  // `-ly` is the strongest suffix signal in English, but "reply" and "apply" are not
  // adverbs, so require something before it that looks like a stem.
  if (w.length > 4 && w.endsWith('ly') && !/[aeiou]ly$/.test(w)) return 'adverb';

  if (w.length > 4 && (w.endsWith('ing') || w.endsWith('ed'))) return 'verb';
  if (w.length > 3 && w.endsWith('s') && !w.endsWith('ss') && !w.endsWith('us')) {
    // Ambiguous between a plural noun and a third-person verb; nouns are far more
    // common in prose, so this is the cheaper mistake.
    return 'noun';
  }
  if (
    w.length > 4 &&
    /(ous|ful|ive|able|ible|al|ic|ish|less|ary)$/.test(w)
  ) {
    return 'adjective';
  }
  if (
    w.length > 4 &&
    /(tion|sion|ness|ment|ity|ance|ence|ship|hood|ist|ism|er|or)$/.test(w)
  ) {
    return 'noun';
  }
  // Anything left that is a plain word is most often a noun in prose.
  if (w.length > 2) return 'noun';
  return null;
}

/** Word ranges over the document, as `[start, end, word]`. */
function words(text: string): Array<[number, number, string]> {
  const out: Array<[number, number, string]> = [];
  const re = /[A-Za-z][A-Za-z'-]*/g;
  let m;
  while ((m = re.exec(text)) !== null) out.push([m.index, m.index + m[0].length, m[0]]);
  return out;
}

export class PartsOfSpeech {
  static LAYER = 'parts-of-speech';

  editor: MarkdownEditor;
  enabled: boolean;
  roles: Record<'noun' | 'verb' | 'adjective' | 'adverb', number>;
  timer: number;
  onChange: () => void;

  /** @param {import('../src/editor.js').MarkdownEditor} editor */
  constructor(editor: MarkdownEditor) {
    this.editor = editor;
    this.enabled = false;
    this.roles = {
      noun: editor.internRole('pos-noun'),
      verb: editor.internRole('pos-verb'),
      adjective: editor.internRole('pos-adjective'),
      adverb: editor.internRole('pos-adverb'),
    };
    this.timer = 0;
    this.onChange = () => this.schedule();
  }

  toggle() {
    if (this.enabled) this.disable();
    else this.enable();
    return this.enabled;
  }

  enable() {
    if (this.enabled) return;
    this.enabled = true;
    this.editor.addEventListener('change', this.onChange);
    this.recompute();
  }

  disable() {
    if (!this.enabled) return;
    this.enabled = false;
    this.editor.removeEventListener('change', this.onChange);
    clearTimeout(this.timer);
    this.editor.clearLayer(PartsOfSpeech.LAYER);
  }

  /**
   * Tagging the whole document on every keystroke would be wasteful, and the core
   * already slides existing spans over an edit so they stay on their words in the
   * meantime (DESIGN §5.3). Coalescing to idle is therefore invisible.
   */
  schedule() {
    if (!this.enabled) return;
    clearTimeout(this.timer);
    this.timer = setTimeout(() => this.recompute(), 150);
  }

  recompute() {
    if (!this.enabled) return;
    const text = this.editor.markdown;
    const spans = [];
    for (const [start, end, word] of words(text)) {
      const tag = tagWord(word);
      if (tag) spans.push({ start, end, role: this.roles[tag] });
    }
    this.editor.setLayer(PartsOfSpeech.LAYER, spans);
  }
}

File: apple/Sources/MDEHost/PartsOfSpeech.swift
import Foundation
import MDECore
import MDEditorUI
import NaturalLanguage

/// Parts-of-speech highlighting — an extension, not a feature of the editor.
///
/// Like `TypewriterMode`, nothing in `MDEditorUI` knows this exists. It reads the
/// document, asks the operating system what each word is, and pushes a decoration layer
/// (DESIGN §5.3).
///
/// It is the more demanding of the two showcase extensions, because its decorations
/// depend on *language* — something the markdown parser has no concept of and never
/// will. The core supplies ranges and identity; what a range *means* is entirely the
/// host's business, which is what makes a feature like this possible without touching
/// the editor.
public final class PartsOfSpeech: MarkdownPlugin {
    public let name = "mde.parts-of-speech"
    private static let layer = "words"
    private static let analysis = "tag-document"

    private var context: MarkdownPluginContext?
    private var roles: [NLTag: UInt32] = [:]
    public private(set) var isEnabled = false

    /// The attributes this extension's roles need. See `TypewriterMode.themeRoles`.
    public static func themeRoles() -> [String: [NSAttributedString.Key: Any]] {
        [
            "pos-noun": [.foregroundColor: PlatformColor.systemBlue],
            "pos-verb": [.foregroundColor: PlatformColor.systemOrange],
            "pos-adjective": [.foregroundColor: PlatformColor.systemPurple],
            "pos-adverb": [.foregroundColor: PlatformColor.systemTeal],
        ]
    }

    public init() {}

    public func install(in context: MarkdownPluginContext) throws {
        self.context = context
        roles = [
            .noun: context.internRole("pos-noun"),
            .verb: context.internRole("pos-verb"),
            .adjective: context.internRole("pos-adjective"),
            .adverb: context.internRole("pos-adverb"),
        ]
    }

    public func uninstall() {
        isEnabled = false
        context = nil
        roles.removeAll()
    }

    public func markdownDidChange() { scheduleRecompute() }

    @discardableResult
    public func toggle() -> Bool {
        isEnabled ? disable() : enable()
        return isEnabled
    }

    public func enable() {
        guard !isEnabled else { return }
        isEnabled = true
        recompute()
    }

    public func disable() {
        guard isEnabled else { return }
        isEnabled = false
        context?.cancelAnalysis(Self.analysis)
        context?.clearLayer(Self.layer)
    }

    /// Tagging the whole document on every keystroke would be wasteful, and the core
    /// already slides existing spans over an edit so they stay on their words in the
    /// meantime (DESIGN §5.3) — so coalescing to a short idle is invisible.
    public func scheduleRecompute() {
        scheduleRecompute(after: 0.15)
    }

    public func recompute() {
        scheduleRecompute(after: 0)
    }

    private func scheduleRecompute(after delay: TimeInterval) {
        guard isEnabled, let context else { return }
        let roles = roles
        context.scheduleAnalysis(
            Self.analysis,
            delay: delay,
            analyze: { text, cancellation in
                Self.tag(text, roles: roles, cancellation: cancellation)
            },
            apply: { spans, context in context.setLayer(Self.layer, spans) }
        )
    }

    private static func tag(
        _ text: String,
        roles: [NLTag: UInt32],
        cancellation: MarkdownPluginAnalysisCancellation
    ) -> [LayerSpan] {
        guard !text.isEmpty else { return [] }

        let tagger = NLTagger(tagSchemes: [.lexicalClass])
        tagger.string = text

        var spans: [LayerSpan] = []
        tagger.enumerateTags(
            in: text.startIndex ..< text.endIndex,
            unit: .word,
            scheme: .lexicalClass,
            options: [.omitWhitespace, .omitPunctuation, .omitOther]
        ) { tag, range in
            guard !cancellation.isCancelled else { return false }
            guard let tag, let role = roles[tag] else { return true }
            // NSRange from String.Index, so the offsets are UTF-16 — which is what every
            // boundary in this API speaks (DESIGN §3.2).
            spans.append(LayerSpan(range: NSRange(range, in: text), role: role))
            return true
        }
        return spans
    }
}
